### PR TITLE
Arm Control Reimplementation

### PR DESCRIPTION
**Pull request by ChielBruin**
# This PR should be merged after #27, as this one includes most of those changes as well

Ported the arm control to Python and converted it to a state machine using the Smach framework:
- Added deprecation warning to the 'old' arm control
- Implemented a basic state machine that follows the diagram created in the WPR assignment
- Created function for arm movement (original PR #17)
- Created functions for IO operations (original PR #26)
- Created functions that add collidable objects to the scene (original PR #27)

The new arm control node is a full functional replacement of the old version, but still needs a lot of work. Many states are still missing (primarily when things go wrong) and state transitions should be added. This can however only be done after specifying a better version of our current state machine with the group.

### CODE REVIEW

File: catkin_ws/src/kpr_arm_control/src/py_arm_control/arm_control.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+											
+		smach.StateMachine.add('MoveToStart', state.MoveToStart(moveArmTo), 
+							   transitions={'MoveOK':'OK',
+											'MoveError':'MOVE_ERR'},
+							   remapping={	'start':'startingPosition'})
+	return sm
+
+def moveArmTo(pose_target):
+'''
+Plan the movement to the secified position and move the arm.
```
Comment (MaxGroenenboom) [inline catkin_ws/src/kpr_arm_control/src/py_arm_control/arm_control.py]: Typo

File: catkin_ws/src/kpr_arm_control/src/py_arm_control/arm_control.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+Set the Value of a specified IO pin.
+Note that it only writes to digital pins as we currently are not using the analog pins.
+
+@param pin: The pin to write to
+@param value: The value that is written to the pin
+@return True when the value is set correctly, False otherwise
+'''
+	try:
+		set_io = rospy.ServiceProxy('set_io', SetIO)
+		#set_io = rospy.ServiceProxy('set_io_testing', SetIO)	# Use these lines when testing using robot.launch
```
Comment (MaxGroenenboom) [inline catkin_ws/src/kpr_arm_control/src/py_arm_control/arm_control.py:147]: Why not use the topic rebind function instead?

File: catkin_ws/src/kpr_arm_control/src/py_arm_control/arm_control.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+	
+def setupIO():
+'''
+Setup the IO subscriber and the setIO service.
+@return The subscriber to the io_states channel
+'''
+	rospy.wait_for_service('set_io')
+	io_states_sub = rospy.Subscriber("io_states", IOStates, IOStatesCallback);
+	#rospy.wait_for_service('set_io_testing')											# Use these lines when testing using robot.launch
+	#io_states_sub = rospy.Subscriber("io_states_testing", IOStates, IOStatesCallback);	# Use these lines when testing using robot.launch
```
Comment (MaxGroenenboom) [inline catkin_ws/src/kpr_arm_control/src/py_arm_control/arm_control.py:181]: Same as above

File: catkin_ws/src/kpr_arm_control/src/py_arm_control/states.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+'''
+State representing the gripping of the cucumber using the vacuum suction cup.
+'''
+	def __init__(self, setIO):
+		self.setIO = setIO
+		smach.State.__init__(self, outcomes=['VacuumCreated', 'VacuumError'])
+
+	def execute(self, userdata):
+	'''
+	Enable the vacuum and make sure it is correctly initialted.
```
Comment (MaxGroenenboom) [inline catkin_ws/src/kpr_arm_control/src/py_arm_control/states.py]: typo!

File: catkin_ws/src/kpr_testing/src/io_pins.py [hunk 5]
```diff
@@ -108,14 +111,16 @@ def get_message():
 	Analog In
 	Analog Out
 	'''
+	global pinStates
+	
 	msg = IOStates()
 	for pin in pinStates:
 		if pinStates[pin][1] == "Digital Out":
-			msg.digital_out_states.append(Digital(int(pin[2]), bool(pinStates[pin][0])))
+			msg.digital_out_states.append(Digital(int(pin[2]), bool(int(float(pinStates[pin][0])))))
```
Comment (MaxGroenenboom) [inline catkin_ws/src/kpr_testing/src/io_pins.py:119]: iew

File: catkin_ws/src/kpr_testing/src/io_pins.py [hunk 6]
```diff
@@ -207,7 +212,8 @@ def main():
 	rospy.loginfo("Started")
 	s = rospy.Service('set_io_testing', SetIO, request)
 	display()
-	io_state_pub = rospy.Publisher("io_states", IOStates, queue_size=10)
+	#io_state_pub = rospy.Publisher("io_states", IOStates, queue_size=10)
```
Comment (MaxGroenenboom) [inline catkin_ws/src/kpr_testing/src/io_pins.py]: Can't this line be removed?

**Comment by MaxGroenenboom** [comment]
It doesn't run for me, I get an `IndentationError`:
```
  File "/home/kaspr/git/tomatenplukkers/catkin_ws/src/kpr_arm_control/src/py_arm_control/arm_control.py", line 37
    '''
      ^
IndentationError: expected an indented block
```
According to stackoverflow the documentation has to be indented as well.

**Comment by ChielBruin** [comment]
I should have tested that!

**Comment by ChielBruin** [comment]
@MaxGroenenboom It should be fixed now

**Review by MaxGroenenboom — APPROVED** [review]

**Pull request merged**